File: backend/internal/domain/guide_about.go
package domain

import (
	"regexp"
	"strings"
	"unicode/utf8"
)

var guideAboutLinkRe = regexp.MustCompile(`(?i)` +
	`\[(?:[^\]]*)\]\(\s*(?:https?://|www\.)[^)]+\)|` +
	`(?:https?://|www\.)[^\s<>"']+|` +
	`\b(?:t\.me|telegram\.me|wa\.me|instagram\.com|facebook\.com|fb\.com|tiktok\.com|youtube\.com|youtu\.be|linkedin\.com|vk\.com|maps\.app\.goo\.gl)(?:/[^\s<>"']*)?|` +
	`\b[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}\b|` +
	`@[A-Za-z][A-Za-z0-9_]{3,31}\b`)

var guideAboutSpaceRe = regexp.MustCompile(`[^\S\n]{2,}`)
var guideAboutNLRe = regexp.MustCompile(`\n{3,}`)

// AcceptGuideAbout keeps existing over-limit values, but rejects growth past GuideAboutMaxLen.
func AcceptGuideAbout(next, current string) (string, error) {
	n := utf8.RuneCountInString(next)
	if n <= GuideAboutMaxLen {
		return next, nil
	}
	if utf8.RuneCountInString(current) >= n {
		return next, nil
	}
	return "", ErrGuideAboutTooLong
}

// PublicGuideAbout is the visitor-facing about text: links and contact handles are hidden.
func PublicGuideAbout(s string) string {
	s = guideAboutLinkRe.ReplaceAllString(s, " ")
	lines := strings.Split(s, "\n")
	for i, line := range lines {
		lines[i] = strings.TrimSpace(guideAboutSpaceRe.ReplaceAllString(line, " "))
	}
	s = strings.Join(lines, "\n")
	s = guideAboutNLRe.ReplaceAllString(s, "\n\n")
	return strings.TrimSpace(s)
}

File: packages/guides/src/lib/guideAbout.ts
export const GUIDE_ABOUT_MAX_LEN = 1000

export function guideAboutLen(text = ''): number {
  return Array.from(text).length
}

export function clampGuideAbout(text: string): string {
  const chars = Array.from(text)
  if (chars.length <= GUIDE_ABOUT_MAX_LEN) return text
  return chars.slice(0, GUIDE_ABOUT_MAX_LEN).join('')
}

export function stripGuideAboutLinks(text: string): string {
  const cleaned = text.replace(
    /\[(?:[^\]]*)\]\(\s*(?:https?:\/\/|www\.)[^)]+\)|(?:https?:\/\/|www\.)[^\s<>"']+|\b(?:t\.me|telegram\.me|wa\.me|instagram\.com|facebook\.com|fb\.com|tiktok\.com|youtube\.com|youtu\.be|linkedin\.com|vk\.com|maps\.app\.goo\.gl)(?:\/[^\s<>"']*)?|\b[A-Z0-9._%+\-]+@[A-Z0-9.\-]+\.[A-Z]{2,}\b|@[A-Za-z][A-Za-z0-9_]{3,31}\b/gi,
    ' ',
  )
  const lines = cleaned.split('\n').map((line) => line.replace(/[^\S\n]{2,}/g, ' ').trim())
  return lines.join('\n').replace(/\n{3,}/g, '\n\n').trim()
}

export function displayGuideAbout(text?: string | null): string {
  if (!text) return ''
  return clampGuideAbout(stripGuideAboutLinks(text))
}

export function nextGuideAbout(current: string, next: string): string {
  const nextLen = guideAboutLen(next)
  const currentLen = guideAboutLen(current)
  if (nextLen <= GUIDE_ABOUT_MAX_LEN) return next
  if (currentLen > GUIDE_ABOUT_MAX_LEN && nextLen < currentLen) return next
  return clampGuideAbout(next)
}
